### PR TITLE
In 152464351 refactor allegations edit

### PR DESCRIPTION
**Pull request by saraOSI**
### Pivotal Story

- [Implement Redux validations and state management for **Allegations** card](https://www.pivotaltracker.com/story/show/152464351)

### Purpose
Redux refactor for allegations edit

### Notes for Reviewer
Screening decision is up next. Two feature specs marked pending until the state of those cards is in the redux store.

Also, I may have fixed cross reports?

### CODE REVIEW

File: spec/javascripts/views/allegations/AllegationsFormSpec.jsx [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+  it('renders "required" if allegations are required', () => {
+    const component = renderAllegationsForm({required: true})
+    expect(component.find('th').at(2).text()).toEqual('Allegation(s) (Required)')
+  })
+
+  it('renders the victim name for an allegation in the first column', () => {
+    const allegations = [{victimName: 'John Smith', allegationTypes: []}]
+    const component = renderAllegationsForm({allegations})
+    expect(component.find('tbody').find('tr').find('td').at(0).text()).toEqual('John Smith')
```
Comment (TylerOSI) [inline spec/javascripts/views/allegations/AllegationsFormSpec.jsx]: not a required change, but i think its possible to combine these finds to be `find('tbody tr td')`. This comment can apply in other places too.

File: app/javascript/containers/screenings/AllegationsFormContainer.jsx [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+import AllegationsForm from 'views/allegations/AllegationsForm'
+import {
+  getFormattedAllegationsSelector,
+} from 'selectors/screening/allegationsFormSelectors'
+import ALLEGATION_TYPES from 'enums/AllegationTypes'
+
+const mapStateToProps = (state, {toggleMode}) => (
+  {
+    allegations: getFormattedAllegationsSelector(state).toJS(),
+    allegationTypes: ALLEGATION_TYPES.map((type) => ({label: type.value, value: type.value})),
```
Comment (TylerOSI) [inline app/javascript/containers/screenings/AllegationsFormContainer.jsx]: This is fairly simple logic, but i think it should be in a selector. You can then test that for a particular value, that it is mapped correctly.

File: app/javascript/selectors/screening/allegationsFormSelectors.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+  (victims, perpetrators, allegations) => (
+    victims.map((victim) => (
+      perpetrators.map((perpetrator, index) => {
+        const victimId = victim.get('id')
+        const perpetratorId = perpetrator.get('id')
+        const allegationTypes = allegations.find((allegation) => (
+          allegation.get('victimId') === victimId && allegation.get('perpetratorId') === perpetratorId
+        ), null, Map()).get('allegationTypes', List())
+        return fromJS({
+          victimName: index === 0 ? nameFormatter(victim.toJS()) : '',
```
Comment (TylerOSI) [inline app/javascript/selectors/screening/allegationsFormSelectors.js]: why only the first perpetrator? Im assuming this is a business rule?

File: spec/javascripts/views/allegations/AllegationsFormSpec.jsx (hunk 1 above)
Comment (saraOSI) [reply spec/javascripts/views/allegations/AllegationsFormSpec.jsx]: Neat! I'll try that.

File: app/javascript/selectors/screening/allegationsFormSelectors.js (hunk 3 above)
Comment (saraOSI) [reply app/javascript/selectors/screening/allegationsFormSelectors.js]: The layout shows the victim name once per group of perpetrators.

File: app/javascript/reducers/allegationsFormReducer.js [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+      const {allegations} = screening
+      return buildAllegations(allegations)
+    }
+  },
+  [SET_ALLEGATION_TYPES](state, {payload: {victimId, perpetratorId, allegationTypes}}) {
+    return state.filterNot((allegation) => (
+      allegation.get('victimId') === victimId && allegation.get('perpetratorId') === perpetratorId
+    )).push(fromJS({id: null, victimId, perpetratorId, allegationTypes}))
+  },
+  [RESET_ALLEGATIONS_FORM](state, {payload: {allegations}}) {
```
[the remark this answers is not in the document]
Comment (saraOSI) [reply app/javascript/reducers/allegationsFormReducer.js]: Good call. I'm surprised lint didn't catch this.

**Pull request merged**